### PR TITLE
Add `cupy.byte_bounds`

### PR DESCRIPTION
**Pull request by maronuu**
Hi everyone,
This PR is following tracker #6078, adding support for `cupy.byte_bounds` API.
It is my first PR towards CuPy. I would appreciate any helpful viewpoints.
Thanks!

## What
Added `cupy.byte_bounds` API.
`byte_bounds` returns pointers of the end-points of an array, which can be calculated from `ptr`, `strides`, `shape`, and `itemsize` attributes.

## QA
unit testing

### CODE REVIEW

File: cupy/_misc/byte_bounds.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    bytes_a = a.dtype.itemsize
+
+    if astrides is None:
+        # contiguous case
+        a_high += a.size * bytes_a
+    else:
+        # non-contiguous case
+        for shape, stride in zip(ashape, astrides):
+            if stride < 0:
+                a_low += (shape-1)*stride
```
Comment (emcastillo) [inline cupy/_misc/byte_bounds.py:23]:
```suggestion
                a_low += (shape - 1) * stride
```
please format operations like this :)

Comment (maronuu) [reply cupy/_misc/byte_bounds.py]: Done in the commit b18562ad747cc29ea5358821a614bfc13bbdea0c.
Thanks!

File: cupy/_misc/byte_bounds.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        Tuple[int, int]: pointers to the end-points of an array
+
+    .. seealso:: :func:`numpy.byte_bounds`
+    """
+    a_low = a_high = a.data.ptr
+    astrides = a.strides
+    ashape = a.shape
+    bytes_a = a.dtype.itemsize
+
+    if astrides is None:
```
[the remark this answers is not in the document]
Comment (maronuu) [reply cupy/_misc/byte_bounds.py]: > `cupy.ndarray.strides` always returns a tuple, not `None`.

I checked the corresponding [code](https://github.com/cupy/cupy/blob/v11.0.0/cupy/_core/core.pyx#L381), and understood that. Implementation in commit c2a3e8849e4b66622b779527f20c99bccc4ed593 supports it.

However, [docs](https://docs.cupy.dev/en/stable/reference/generated/cupy.ndarray.html#cupy.ndarray) says that 
> strides (tuple of ints or [None](https://docs.python.org/3/library/constants.html#None)) – Strides of data in memory.

Is it a mistake in API reference? If so, should I include the correction of this reference in this PR?

File: tests/cupy_tests/misc_tests/test_byte_bounds.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import unittest
+
+import cupy
+from cupy import testing
+
+
+@testing.gpu
+class TestByteBounds(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (maronuu) [reply tests/cupy_tests/misc_tests/test_byte_bounds.py]: commit: 3d07223e55928068f30675960804777d9184d0f3

Add some cases of
- negative strides 1D
- negative strides 2D
- positive and negative strides (mixed) 2D
- many dtypes (built-in numeric types, `cupy.int`, `cupy.uint`, `cupy.float`, `cupy.complex`)

File: cupy/_misc/byte_bounds.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/_misc/byte_bounds.py]: If `strides=None` is passed, `ndarray.strides` is inferred from shape and itemsize here.
https://github.com/cupy/cupy/blob/master/cupy/_core/core.pyx#L216

File: tests/cupy_tests/misc_tests/test_byte_bounds.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (asi1024) [reply tests/cupy_tests/misc_tests/test_byte_bounds.py]: How about using `testing.for_all_dtypes` decorator to parametrize dtypes?

https://docs.cupy.dev/en/stable/reference/generated/cupy.testing.for_all_dtypes.html

[the remark this answers is not in the document]
Comment (maronuu) [reply tests/cupy_tests/misc_tests/test_byte_bounds.py]: commit: 145f775fde192aed4988f7a19c37187fe3c17680

uses `testing.for_all_dtypes`.

**Review by asi1024 — APPROVED** [review]

**Comment by asi1024** [comment]
LGTM. Thanks for your contribution!

**Pull request merged**